### ISSUE TITLE
Unexpected EOF when trying to start non-existing container

### ISSUE THREAD

**Issue by tejksat** [docker/for-mac#5025]
- [x] I have tried with the latest version of my channel (Stable)
  - [x] I have uploaded Diagnostics
  - Diagnostics ID: 2BE476CF-AF05-44A9-A95A-DE9A45DB43A6/20201028155816

### Description

When trying to start a non-existing Docker container daemon does not reply with `404` as it is declared in [Docker Engine API Reference](https://docs.docker.com/engine/api/v1.40/#operation/ContainerStart) but it rather unexpectedly interrupts the connection.

### Expected behavior

```
$ curl -v --unix-socket /var/run/docker.sock -X POST http://localhost/v1.40/containers/non-existing/start
*   Trying /var/run/docker.sock:0...
* Connected to localhost (docker.sock) port 80 (#0)
> POST /v1.40/containers/non-existing/start HTTP/1.1
> Host: localhost
> User-Agent: curl/7.71.1
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 404 Not Found
< Api-Version: 1.40
< Content-Type: application/json
< Date: Wed, 28 Oct 2020 16:08:31 GMT
< Docker-Experimental: false
< Ostype: linux
< Server: Docker/19.03.13 (linux)
< Transfer-Encoding: chunked
<
{"message":"No such container: non-existing"}
* Connection #0 to host localhost left intact
```

_or_

```
$ docker start non-existing
Error response from daemon: No such container: non-existing
```

### Actual behavior

```
$ curl -v --unix-socket /var/run/docker.sock -X POST http://localhost/v1.40/containers/non-existing/start
*   Trying /var/run/docker.sock:0...
* Connected to localhost (docker.sock) port 80 (#0)
> POST /v1.40/containers/non-existing/start HTTP/1.1
> Host: localhost
> User-Agent: curl/7.71.1
> Accept: */*
>
* Empty reply from server
* Connection #0 to host localhost left intact
curl: (52) Empty reply from server
```

_or_

```
$ docker start non-existing
error during connect: Post http://%2Fvar%2Frun%2Fdocker.sock/v1.40/containers/non-existing/start: EOF
Error: failed to start containers: non-existing
```

### Information

  - macOS Version: 10.15.7 (macOS Catalina)

### Diagnostic logs

```
Diagnose succeeded
```

### Steps to reproduce the behavior

1. Run `curl -v --unix-socket /var/run/docker.sock -X POST http://localhost/v1.40/containers/non-existing/start` in Terminal.
_or_
2. Run `docker start non-existing` in Terminal.

**Comment by thaJeztah** [comment]
@djs55 ptal

**Comment by gesellix** [comment]
Is this one related to #5058?

**Comment by thaJeztah** [comment]
yes, looks the same as #5058 and #5059

**Comment by stephen-turner** [comment]
Thanks, we're on it.

**Comment by mat007** [comment]
We believe this is now fixed in Edge 2.5.1.0, I’m going to close this issue but feel free to comment if it doesn’t solve it for you.
This will reach Stable most likely mid-December.

**Comment by thaJeztah** [comment]
@mat007 looks like it's not resolved; on docker desktop edge `2.5.1.0 (49923)`;

Incorrect response with the proxied socket;

```console
$ curl -v --unix-socket /var/run/docker.sock http://localhost/containers/nosuchcontainer/json

*   Trying /var/run/docker.sock:0...
* Connected to localhost (docker.sock) port 80 (#0)
> GET /containers/nosuchcontainer/json HTTP/1.1
> Host: localhost
> User-Agent: curl/7.65.3
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 404 Not Found
< Api-Version: 1.41
< Content-Type: application/json
< Date: Fri, 20 Nov 2020 14:23:21 GMT
< Docker-Experimental: true
< Ostype: linux
< Server: Docker/20.10.0-rc1 (linux)
< Transfer-Encoding: chunked
<
* transfer closed with outstanding read data remaining
* Closing connection 0
curl: (18) transfer closed with outstanding read data remaining
```

Works correct with the 'raw' socket:

```console
$ curl -v --unix-socket ~/Library/Containers/com.docker.docker/Data/docker.raw.sock http://localhost/containers/nosuchcontainer/json

*   Trying /Users/sebastiaan/Library/Containers/com.dock:0...
* Connected to localhost (/Users/sebastiaan/Library/Containers/com.dock) port 80 (#0)
> GET /containers/nosuchcontainer/json HTTP/1.1
> Host: localhost
> User-Agent: curl/7.65.3
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 404 Not Found
< Api-Version: 1.41
< Content-Type: application/json
< Docker-Experimental: true
< Ostype: linux
< Server: Docker/20.10.0-rc1 (linux)
< Date: Fri, 20 Nov 2020 14:22:40 GMT
< Content-Length: 49
<
{"message":"No such container: nosuchcontainer"}
* Connection #0 to host localhost left intact
```

**Comment by tejksat** [comment]
I could add that Docker CLI output of the command also contains a questionable message in the first line:
```
$ docker start non-existing
Error reading JSON: invalid character 'u' looking for beginning of value
Error: failed to start containers: non-existing
```

**Comment by thaJeztah** [comment]
Ah, yes, and in that case also works if the raw socket is used;

```console
$ docker -H unix://"$HOME"/Library/Containers/com.docker.docker/Data/docker.raw.sock start nosuchcontainer
Error response from daemon: No such container: nosuchcontainer
Error: failed to start containers: nosuchcontainer
```

**Comment by simonferquel** [comment]
@thaJeztah can you send a diagnostic ? I fixed a panic triggering the bug on Windows and WSL, but there might be a specific mac issue here. (should be easy enough to diagnostic though as the proxy logs stack traces on panics)

**Comment by thaJeztah** [comment]
@simonferquel `4339D5BE-D560-46F3-A796-F7E6EF442494/20201120154119`

**Comment by simonferquel** [comment]
Really useful, thanks:
```
goroutine 937 [running]:
github.com/docker/pinata/common/pkg/proxy.(*GuiAnalytics).sendCommand(0xc00041ac40, 0x54e6102, 0xe, 0x54dbd36, 0x7, 0x1)
	/Users/administrator/jenkins/workspace/sktop_desktop-build_edge-2.5.1.x/src/github.com/docker/pinata/common/pkg/proxy/analytics.go:112 +0x2af
github.com/docker/pinata/common/pkg/proxy.(*GuiAnalytics).trackRouteForHTTPMethod(0xc00041ac40, 0xc000884d00, 0x613d2b8, 0x5597bb0)
	/Users/administrator/jenkins/workspace/sktop_desktop-build_edge-2.5.1.x/src/github.com/docker/pinata/common/pkg/proxy/analytics.go:98 +0x145
github.com/docker/pinata/common/pkg/proxy.(*GuiAnalytics).Middleware.func1(0x56e98e0, 0xc000744540, 0xc000884d00)
	/Users/administrator/jenkins/workspace/sktop_desktop-build_edge-2.5.1.x/src/github.com/docker/pinata/common/pkg/proxy/analytics.go:60 +0x7c
net/http.HandlerFunc.ServeHTTP(0xc0006fd4c0, 0x56e98e0, 0xc000744540, 0xc000884d00)
	/Users/administrator/.pinata/downloads/go/1.15.2/go/src/net/http/server.go:2042 +0x44
github.com/gorilla/mux.(*Router).ServeHTTP(0xc00039e540, 0x56e98e0, 0xc000744540, 0xc000884b00)
	/Users/administrator/jenkins/workspace/sktop_desktop-build_edge-2.5.1.x/src/github.com/docker/pinata/vendor/github.com/gorilla/mux/mux.go:210 +0xd3
net/http.serverHandler.ServeHTTP(0xc0001900e0, 0x56e98e0, 0xc000744540, 0xc000884b00)
	/Users/administrator/.pinata/downloads/go/1.15.2/go/src/net/http/server.go:2843 +0xa3
net/http.(*conn).serve(0xc000440960, 0x56ec520, 0xc000142dc0)
	/Users/administrator/.pinata/downloads/go/1.15.2/go/src/net/http/server.go:1925 +0x8ad
created by net/http.(*Server).Serve
	/Users/administrator/.pinata/downloads/go/1.15.2/go/src/net/http/server.go:2969 +0x36c
```

**Comment by simonferquel** [comment]
Oh actually the problem exist on Windows as well. I was confused because previously the client would not receive response at all (not even the headers). Now the headers are correctly received, but the body in Chunked encoding is not:
```
 curl -v --unix-socket /var/run/docker.sock http://localhost/containers/nosuchcontainer/json
*   Trying /var/run/docker.sock:0...
* Connected to localhost (/var/run/docker.sock) port 80 (#0)
> GET /containers/nosuchcontainer/json HTTP/1.1
> Host: localhost
> User-Agent: curl/7.68.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 404 Not Found
< Api-Version: 1.41
< Content-Type: application/json
< Date: Fri, 20 Nov 2020 16:10:11 GMT
< Docker-Experimental: true
< Ostype: linux
< Server: Docker/20.10.0-beta1 (linux)
< Transfer-Encoding: chunked
<
* transfer closed with outstanding read data remaining
* Closing connection 0
curl: (18) transfer closed with outstanding read data remaining
```

**Comment by seanhamlin** [comment]
Also seeing this issue I believe (Mac OSX 10.15.7). I updated to docker edge (`Docker version 20.10.0-rc1, build 5cc2396`)

```bash
$ docker-compose build
Traceback (most recent call last):
  File "site-packages/urllib3/response.py", line 696, in _update_chunk_length
ValueError: invalid literal for int() with base 16: b''

During handling of the above exception, another exception occurred:
...
Traceback (most recent call last):
  File "docker-compose", line 3, in <module>
  File "compose/cli/main.py", line 67, in main
  File "compose/cli/main.py", line 123, in perform_command
  File "compose/cli/command.py", line 69, in project_from_options
  File "compose/cli/command.py", line 141, in get_project
  File "compose/project.py", line 117, in from_config
  File "compose/project.py", line 957, in get_volumes_from
  File "compose/project.py", line 957, in <listcomp>
  File "compose/project.py", line 946, in build_volume_from
  File "compose/container.py", line 44, in from_id
  File "site-packages/docker/utils/decorators.py", line 19, in wrapped
  File "site-packages/docker/api/container.py", line 771, in inspect_container
  File "site-packages/docker/utils/decorators.py", line 46, in inner
  File "site-packages/docker/api/client.py", line 228, in _get
  File "site-packages/requests/sessions.py", line 543, in get
  File "site-packages/requests/sessions.py", line 530, in request
  File "site-packages/requests/sessions.py", line 685, in send
  File "site-packages/requests/models.py", line 829, in content
  File "site-packages/requests/models.py", line 754, in generate
requests.exceptions.ChunkedEncodingError: ('Connection broken: IncompleteRead(0 bytes read)', IncompleteRead(0 bytes read))
[85343] Failed to execute script docker-compose
```

and

```bash
$ docker start non-existing
Error reading JSON: invalid character 'c' looking for beginning of value
Error: failed to start containers: non-existing
```

```bash
$ docker -H unix://"$HOME"/Library/Containers/com.docker.docker/Data/docker.raw.sock start nosuchcontainer
Error response from daemon: No such container: nosuchcontainer
Error: failed to start containers: nosuchcontainer
```

Diagnostics id `125481F4-D3A7-40DF-B851-2A56C665C2F9/20201122211313`

**Comment by mat007** [comment]
This should really be fixed in all places in the code this time in Edge 2.5.1.0 (49923).
I’m closing this issue, but if you’re still seeing this after updating to that version please keep comment again.
Thanks!

**Comment by mat007** [comment]
Sorry I jumped the gun a bit, that version is due to be released today. 😁

**Comment by mat007** [comment]
It’s just been released in Edge 2.5.2 (50165).
Thank you for your patience!
